Pin xoxo unpin

aliyss opened the pull request:
chat.pin()
chat.unpin()

client.pinChat(chatId)
...

aliyss committed:
Merge pull request #8 from pedroslopez/master

Merge To Updated Version

aliyss committed:
Merge pull request #9 from pedroslopez/master

Merge

aliyss committed:
Merge pull request #10 from pedroslopez/master

Merge

aliyss committed:
Added Pinning
- chat.pin() to pin
- chat.unpin() to unpin

aliyss committed:
Added to example.js

pedroslopez reviewed (changes requested):
Pinning/unpinning seems to work, but the return value isn't as expected (docstrings indicate it returns a boolean with the pin state). While testing I found pinning/unpinning returns a number that seems like it may be a timestamp, and I even got `undefined` at one point. Can you verify this?

aliyss committed:
Merge branch 'master' into pin-xoxo-unpin

aliyss committed:
Fixed the stuff concerning pins
- Now return Booleans.

aliyss committed:
Merge remote-tracking branch 'origin/pin-xoxo-unpin' into pin-xoxo-unpin

aliyss committed:
Fixed the stuff concerning pins
- Now return Booleans.

pedroslopez reviewed (changes requested):
The return still doesn't seem to be working right. When I first pin a chat I get `false`, and unpinning gives me `true`, which would seem backwards. 

Calling `pin()` or `unpin()` more than once should also be limited or simply ignored. Doing this more than once seems to bug it out and it can get to a state where trying to unpin gives the message "Couldn't unpin chat", and just stays stuck until a restart of whatsapp web. Furthermore, doing this reverts the return values, causing `pin()` to return `true` and `unpin()` to return `false`.

Aliyss commented:
Ok. Will add a check later. So the state is checked before it tries to pin/unpin.

pedroslopez committed:
Merge branch 'master' into pin-xoxo-unpin

pedroslopez commented:
@Aliyss have you been able to take a look at this? Still seems like the pin state isn't too reliable

Aliyss commented:
@pedroslopez I didn't get to it. Very Sorry. Currently a lot to do.

aliyss committed:
Merge branch 'master' into pin-xoxo-unpin

aliyss committed:
Merge branch 'master' into pin-xoxo-unpin

aliyss committed:
Fixed pins
- Added Check for already pinned / unpinned.
- Added Check for already max amount pinned.
- returns true / false based on checks instead of WhatsApp, since Whatsapp needs time to update the value.

aliyss committed:
Merge branch 'master' into pin-xoxo-unpin

aliyss committed:
Fixed Pin and Unpin
- Checks are made in correct order

pedroslopez committed:
Merge branch 'master' into pin-xoxo-unpin

pedroslopez commented on src/Client.js:
```diff
@@ -563,6 +563,40 @@ class Client extends EventEmitter {
+            if (chat.pin) {
+                return true;
+            }
+            let maxPinned = await window.Store.Chat.models[2].pin;
```
Could there be a better way to get the max number of pinned chats allowed / current number of pinned chats for this check?

This seems like it could be error prone if maybe there are less than 3 chats

Aliyss replied:
Yeah... I searched quite a bit. I'll try again with a fresh mind. But couldn't find anything promising.

aliyss committed:
Merge branch 'master' into pin-xoxo-unpin

aliyss committed:
Merge branch 'master' into pin-xoxo-unpin

Aliyss commented:
@pedroslopez I couldn't find anything. I've been searching for quite a while, but there's no clear variable.

pedroslopez commented:
@Aliyss I think it's ok to assume it's 3 as long as we clearly indicate what that number means (e.g `const MAX_PIN_COUNT = 3` or something similar)

However, we still need to get the current number of pinned chats and compare it to prevent the issue.

aliyss committed:
Update Client.js

aliyss committed:
Merge branch 'master' into pin-xoxo-unpin

aliyss committed:
Chat Pinning according to changes requested

pedroslopez commented on src/Client.js:
```diff
@@ -610,6 +610,43 @@ class Client extends EventEmitter {
+     * @returns {boolean}
+     */
+    async pinChat(chatId) {
+        return await this.pupPage.evaluate(async chatId => {
```
```suggestion
        return this.pupPage.evaluate(async chatId => {
```

https://eslint.org/docs/rules/no-return-await I'm going to enable this rule soon

pedroslopez commented on src/Client.js:
```diff
@@ -610,6 +610,43 @@ class Client extends EventEmitter {
+            }
+            const MAX_PIN_COUNT = 3;
+            if (window.Store.Chat.models.length > MAX_PIN_COUNT) {
+                let maxPinned = await window.Store.Chat.models[MAX_PIN_COUNT - 1].pin;
```
```suggestion
                let maxPinned = window.Store.Chat.models[MAX_PIN_COUNT - 1].pin;
```

No need for `await` here.

pedroslopez commented on src/Client.js:
```diff
@@ -610,6 +610,43 @@ class Client extends EventEmitter {
+     */
+    async unpinChat(chatId) {
+        return await this.pupPage.evaluate(async chatId => {
+            let chat = await window.Store.Chat.get(chatId);
```
```suggestion
            let chat = window.Store.Chat.get(chatId);
```

No need for `await` here.

pedroslopez commented on src/Client.js:
```diff
@@ -610,6 +610,43 @@ class Client extends EventEmitter {
+     */
+    async pinChat(chatId) {
+        return await this.pupPage.evaluate(async chatId => {
+            let chat = await window.Store.Chat.get(chatId);
```
```suggestion
            let chat = window.Store.Chat.get(chatId);
```

No need for `await` here. Getting a chat does not return a promise.

pedroslopez commented on src/Client.js:
```diff
@@ -610,6 +610,43 @@ class Client extends EventEmitter {
+     * @returns {boolean}
+     */
+    async unpinChat(chatId) {
+        return await this.pupPage.evaluate(async chatId => {
```
```suggestion
        return this.pupPage.evaluate(async chatId => {
```

pedroslopez reviewed (changes requested):
Just a couple details with `await` usage

aliyss committed:
Update src/Client.js

Co-authored-by: Pedro S. Lopez <pedroslopez@me.com>

aliyss committed:
Update src/Client.js

Co-authored-by: Pedro S. Lopez <pedroslopez@me.com>

aliyss committed:
Update src/Client.js

Co-authored-by: Pedro S. Lopez <pedroslopez@me.com>

aliyss committed:
Update src/Client.js

Co-authored-by: Pedro S. Lopez <pedroslopez@me.com>

aliyss committed:
Update src/Client.js

Co-authored-by: Pedro S. Lopez <pedroslopez@me.com>

Aliyss commented:
Fixed and tested. Still works.

Aliyss commented:
Keeping it open... Until officially closed? All requested changes have been made.

pedroslopez committed:
Merge branch 'master' into pin-xoxo-unpin

pedroslopez committed:
specify return type

pedroslopez committed:
Make docs consistent with the functions on the Chat model

pedroslopez committed:
Update Chat.js

pedroslopez committed:
add typings
